### PR TITLE
BungieNet provider

### PR DESCRIPTION
**Pull request by serber**
Fidler logs

Retrieve access_token

```
POST https://www.bungie.net/platform/app/oauth/token HTTP/1.1
Host: www.bungie.net
User-Agent: Microsoft ASP.NET Core OAuth handler
traceparent: 00-be2254e2f6cdfcc5979241a28582e500-a5400f316989697d-00
Content-Type: application/x-www-form-urlencoded
Content-Length: 204

redirect_uri=***&client_id=***&client_secret=***&code=***&grant_type=authorization_code


HTTP/1.1 200 OK
Date: Thu, 27 Jan 2022 19:19:24 GMT
Content-Type: application/json; charset=utf-8
Connection: keep-alive
Cache-Control: private
Vary: Origin,Accept-Encoding
X-BungieNext-MID2: 128
X-BungieNext-Renderer: Frog Blast the Ventcore
X-UA-Compatible: IE=edge
Access-Control-Allow-Origin: https://www.bungie.net
X-SelfUrl: https://www.bungie.net/platform/app/oauth/token/
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
CF-Cache-Status: DYNAMIC
Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
Server: cloudflare
CF-RAY: 6d44551a2fb21634-DME
Content-Length: 854

{"access_token":"***","token_type":"Bearer","expires_in":3600,"refresh_token":"***","refresh_expires_in":7776000,"membership_id":"***"}
```

Getting user information

```
GET https://www.bungie.net/Platform/User/GetBungieNetUserById/*** HTTP/1.1
Host: www.bungie.net
Accept: application/json
Authorization: Bearer ***
X-API-Key: ***
User-Agent: Microsoft ASP.NET Core OAuth handler
traceparent: 00-be2254e2f6cdfcc5979241a28582e500-f809daa34f7292bc-00


HTTP/1.1 200 OK
Date: Thu, 27 Jan 2022 19:19:25 GMT
Content-Type: application/json; charset=utf-8
Connection: keep-alive
Cache-Control: no-cache
Pragma: no-cache
Expires: -1
Vary: X-Requested-With,Origin,Accept-Encoding
X-BungieNext-MID2: 128
X-BungieNext-Renderer: Frog Blast the Ventcore
X-UA-Compatible: IE=edge
X-Ventcore-Status: 1
Access-Control-Allow-Origin: https://www.bungie.net
Set-Cookie: bunglefrogblastventcore=; expires=Fri, 27-Jan-2017 19:19:25 GMT; path=/; HttpOnly
X-SelfUrl: https://www.bungie.net/Platform/User/GetBungieNetUserById/***/
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
CF-Cache-Status: DYNAMIC
Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
Server: cloudflare
CF-RAY: 6d44551dc9311634-DME
Content-Length: 763

{"Response":{"membershipId":"***","uniqueName":"***#259","displayName":"***","profilePicture":0,"profileTheme":0,"userTitle":0,"successMessageFlags":"0","isDeleted":false,"about":"","firstAccess":"2022-01-27T17:42:22.291Z","lastUpdate":"2022-01-27T17:42:22.291Z","showActivity":true,"locale":"en","localeInheritDefault":false,"showGroupMessaging":true,"profilePicturePath":"/img/profile/avatars/default_avatar.gif","profileThemeName":"d2cover","userTitleDisplay":"Newbie","statusText":"","statusDate":"0001-01-01T00:00:00Z","steamDisplayName":"БОТ Xavier","cachedBungieGlobalDisplayName":"","cachedBungieGlobalDisplayNameCode":259},"ErrorCode":1,"ThrottleSeconds":0,"ErrorStatus":"Success","Message":"Ok","MessageData":{}}
```

### CODE REVIEW

File: src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationConstants.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+/// <summary>
+/// Contains constants specific to the <see cref="BungieNetAuthenticationHandler"/>.
+/// </summary>
+public static class BungieNetAuthenticationConstants
+{
+    public static class Claims
+    {
+        public const string ProfilePicturePath = "profilePicturePath";
+        public const string UniqueName = "uniqueName";
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationConstants.cs:17]: Suggest naming these like the other custom Claim names we have ([examples](https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/search?q=urn)) to start with `urn:`.

File: src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationDefaults.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    /// <summary>
+    /// Default value for <see cref="OAuthOptions.TokenEndpoint"/>.
+    /// </summary>
+    public static readonly string TokenEndpoint = "https://www.bungie.net/platform/app/oauth/token";
+
+    /// <summary>
+    /// Default value for <see cref="OAuthOptions.UserInformationEndpoint"/>.
+    /// </summary>
+    public static readonly string UserInformationEndpoint = "https://www.bungie.net/Platform/User/GetBungieNetUserById/{0}";
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationDefaults.cs:47]: If you look at some of the other providers, in those cases we name this with a `Format` suffix to make it clearer that it isn't a usable value by itself.

File: src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using AspNet.Security.OAuth.BungieNet;
+
+namespace Microsoft.Extensions.DependencyInjection;
+
+/// <summary>
+/// Extension methods to add  authentication capabilities to an HTTP application pipeline.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationExtensions.cs:12]:
```suggestion
/// Extension methods to add BungieNet authentication capabilities to an HTTP application pipeline.
```

File: src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    {
+    }
+
+    protected override async Task<AuthenticationTicket> CreateTicketAsync(
+        [NotNull] ClaimsIdentity identity,
+        [NotNull] AuthenticationProperties properties,
+        [NotNull] OAuthTokenResponse tokens)
+    {
+        var membershipId = tokens.Response?.RootElement.GetString("membership_id");
+        var userInfoEndpoint = string.Format(CultureInfo.InvariantCulture, Options.UserInformationEndpoint, membershipId);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationHandler.cs:35]: Does the membership ID need to be URI-escaped?

File: src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationHandler.cs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        if (!response.IsSuccessStatusCode)
+        {
+            await Log.ExchangeCodeErrorAsync(Logger, response, Context.RequestAborted);
+            return OAuthTokenResponse.Failed(new Exception("An error occurred while retrieving an access token."));
+        }
+
+        var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync(Context.RequestAborted));
+
+        return OAuthTokenResponse.Success(payload);
+    }
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationHandler.cs:92]: It's not obvious to me what functionality this has that isn't the same as the implementation in [the base method](https://github.com/dotnet/aspnetcore/blob/15fa3ad10859abcc54e3ad5557dc928f6c94994d/src/Security/Authentication/OAuth/src/OAuthHandler.cs#L189-L221).

Is this override actually needed?

File: src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationOptions.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+public class BungieNetAuthenticationOptions : OAuthOptions
+{
+    public BungieNetAuthenticationOptions()
+    {
+        ClaimsIssuer = BungieNetAuthenticationDefaults.Issuer;
+        CallbackPath = BungieNetAuthenticationDefaults.CallbackPath;
+
+        AuthorizationEndpoint = BungieNetAuthenticationDefaults.AuthorizationEndpoint;
+        TokenEndpoint = BungieNetAuthenticationDefaults.TokenEndpoint;
+        UserInformationEndpoint = BungieNetAuthenticationDefaults.UserInformationEndpoint;
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationOptions.cs:24]: It doesn't feel right to me to assign this in this way, as it's not valid in of itself. If it's needed to make it work (e.g. to pass validation) then OK, but it just seems a bit of a code smell.

File: src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationHandler.cs (hunk 4 above)
Comment (serber) [reply src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationHandler.cs:35]: No, its number

File: src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationOptions.cs (hunk 6 above)
Comment (serber) [reply src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationOptions.cs:24]: Are you about passing format string into UserInformationEndpoint?

Comment (martincostello) [reply src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationOptions.cs:24]: Yes.

Comment (serber) [reply src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationOptions.cs:24]: What if we change like this
```
public static class BungieNetAuthenticationDefaults
{
...
    public static readonly string UserInformationEndpoint = "https://www.bungie.net/Platform/User/GetBungieNetUserById";
...
}
```

And add membership_id in `BungieNetAuthenticationHandler.CreateTicketAsync` method with string concatenation?

Comment (martincostello) [reply src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationOptions.cs:24]: That would also work - the main thing I'm not a fan of is the fact we need to set a URL that doesn't work by itself and needs per-request mutation.

I'm sure we have another provider somewhere that has this problem, but there's so many I can't remember which one. If I'm right and you can find it, then let's just follow the same approach as that one.

File: src/AspNet.Security.OAuth.BungieNet/AspNet.Security.OAuth.BungieNet.csproj [hunk 7]
```diff
@@ -0,0 +1,24 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <PackageValidationBaselineVersion>6.0.4</PackageValidationBaselineVersion>
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.BungieNet/AspNet.Security.OAuth.BungieNet.csproj:4]:
```suggestion
    <PackageValidationBaselineVersion>6.0.5</PackageValidationBaselineVersion>
```

File: src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationOptions.cs (hunk 6 above)
Comment (serber) [reply src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationOptions.cs:24]: I found few providers with user-specific endpoints

1. [SuperOffice](https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/src/AspNet.Security.OAuth.SuperOffice/SuperOfficeAuthenticationHandler.cs#L47) handler uses string formatting. UserInformationEndpoint [contains](https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/src/AspNet.Security.OAuth.SuperOffice/SuperOfficeAuthenticationOptions.cs#L164) string formatting pattern.
2. [Nextcloud](https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/src/AspNet.Security.OAuth.Nextcloud/NextcloudAuthenticationHandler.cs#L34) handler uses string concatenation.
3. [WorkWeixin](https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/src/AspNet.Security.OAuth.WorkWeixin/WorkWeixinAuthenticationHandler.cs#L49) handler adds userid into query string.

Comment (martincostello) [reply src/AspNet.Security.OAuth.BungieNet/BungieNetAuthenticationOptions.cs:24]: OK, let's just leave this as is then so it's like the SuperOffice implementation.

**Comment by martincostello** [comment]
Closing as this PR has been stale for over a year now without any of the outstanding feedback being addressed.

Please open a new pull request incorporating the requested changes if you wish to come back to this work in the future.

**Pull request closed without merging**